### PR TITLE
Add pickle support for Keras model

### PR DESCRIPTION
**Pull request by LukeWood**
https://github.com/keras-team/keras/issues/19554

Implements some basic unit tests for pickling on top of one of the existing parameterized tests.  I think this gets pretty good coverage across cases.

There might be some weird edge cases where reloading fails, but I think this is fine in that pickling isn't really useful as  a long term storage solution - just a short term way to transfer some info between workers.

Open to ideas about better approaches here - this is just the best I could come up with.

### CODE REVIEW

File: keras/src/models/model_test.py [hunk 1]
```diff
@@ -171,6 +173,18 @@ def test_functional_single_output(self, model_fn, loss_type):
         )
         self.assertListEqual(hist_keys, ref_keys)
 
+        reloaded_pickle = pickle.loads(pickle.dumps(model))
+
+        pred_reloaded = reloaded_pickle.predict(x)
+        pred = model.predict(x)
+
+        # self.assertAllClose fails for some dtypes, so we use np
```
Comment (fchollet) [inline keras/src/models/model_test.py:181]: Is it just a matter of calling `convert_to_numpy()` first? What's the problem?

File: keras/src/models/model_test.py [hunk 2]
```diff
@@ -171,6 +173,18 @@ def test_functional_single_output(self, model_fn, loss_type):
         )
         self.assertListEqual(hist_keys, ref_keys)
 
+        reloaded_pickle = pickle.loads(pickle.dumps(model))
```
Comment (fchollet) [inline keras/src/models/model_test.py:176]: Please add a separate unit test rather than inlining pickle testing here

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/models/model_test.py (hunk 2 above)
Comment (LukeWood) [reply keras/src/models/model_test.py]: good point - added a unique test to cover this

File: keras/src/models/model_test.py (hunk 1 above)
Comment (LukeWood) [reply keras/src/models/model_test.py]: updated to call np.array({tensors}) and this works

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Comment by LukeWood** [comment]
> LGTM, thanks!

awesome - thanks for the help!  Very happy with keras-3!
